### ISSUE TITLE
Constraining to atmospheric CO2 - error

### ISSUE THREAD

**Issue by cahartin** [JGCRI/hector#122]
Error message: "invalid message type in sendMessage". 

There are 3 types of sendMessage
GETDATA, SETDATA, DUMP_TO_DEEP_OCEAN

message == M_DUMP_TO_DEEP_OCEAN in core.cpp  is potentially missing from the core.

However, in core.cpp we only have GET and SET DATA, we are missing the DUMP_TO_DEEP_OCEAN.  This allows us to constrain atmospheric CO2 to a particular input file.

**Comment by rplzzz** [comment]
I wasn't aware that `DUMP_TO_DEEP_OCEAN` is a thing that exists.  Do we expect to use this in normal Hector operations, or is this just a mechanism for imposing constraints?  If the latter, then I think we should consider implementing this sort of thing through a new component ("ConstraintManager" or something like) that reacts as necessary to enforce constraints.  Doing it that way would allow us to configure the constraints more naturally in our input files, and it would allow us to customize the reporting from the component that is implementing the constraints.  It would also be a lot easier to extend through subclassing, if we someday decide we want to create types of constraints that are subtly different from one another.

**Comment by cahartin** [comment]
the idea is that if we running with CO2 concentrations (what we are calling constrained), than any excess carbon is dumped into the deep ocean or carbon is pulled from the deep ocean.  this way we aren't loosing carbon along the way.  

I like the idea of a new component that could deal with all constraints.  Right now we Hector can constrain to CO2, RF, temperature and ocean flux.

**Comment by rplzzz** [comment]
Understood.  So, as I read the code, the carbon forced into the deep ocean is allowed to participate in the ocean part of the carbon cycle (vice being completely sequestered, which is what I had assumed we were doing).  If that's the way we're handling the constraints, then the carbon moved due to constraints has to find its way in or out of the ocean one way or another, irrespective of where the constraints are implemented.   However, even in this case I don't see why we need a special message type just for these deep ocean fluxes.  Why couldn't we just use a `SETDATA` message with a variable name of "to_deep" or something like that?  

My concern here is that I think we should avoid adding any unnecessary complications to the core.  Right now there are two ways that components can interact with one another.  They can send information to another component, or they can request information from it.  We should think carefully about whether we really want to add different, specialized interaction modes.

**Comment by cahartin** [comment]
That makes sense not to add more complications to the core.  At some point we had the capability to constrain to CO2 conc.  Now when I try to implement that feature I get an error. "invalid message type in sendMessage".  

Simple Nbox ln 591 core->sendMessage( M_DUMP_TO_DEEP_OCEAN, D_OCEAN_C, message_data( residual ) );     // update ocean pool

My thought was that the core was missing that send message type.  I should actually change my issue to a question.

**Comment by bpbond** [comment]
I'm confused--when and how did this break? I don't have good enough Internet access to investigate this at the moment, but can do so next week. My memory is that we first used  M_DUMP_TO_DEEP_OCEAN but then realized exactly what Robert notes above, that it would be better to use the general getData/setData mechanism rather than custom messages. It seems that at some point `simpleNbox` got reverted to the previous mechanism, however, and is thus attempting to use M_DUMP_TO_DEEP_OCEAN. Weird.

@rplzzz : to clarify - we used the deep ocean as a 'carbon bank', depositing/withdrawing as necessary, because its turnover time is so slow that it doesn't affect the rest of the system on short time scales. When we want to constrain atmospheric C to an observational record, the Hector carbon cycle does its computational thing, and ends up with a number that's (probably) different from the record. We then adjust the atmospheric pool to match the record, storing the surplus/deficit in the deep ocean to preserve the entire system's mass balance.

**Comment by rplzzz** [comment]
@bpbond I understand.  I was just thinking that if CO<sub>2</sub> (or anything else, for that matter) were being added to or removed from the atmosphere for the purpose of meeting constraints, you would do it by having a completely inert pool dedicated to that purpose, rather than putting it into the deep ocean pool.  If the deep ocean pool interacts slowly enough, then it doesn't make much difference.  Having a dedicated pool would make it easier to report on how much CO<sub>2</sub> was added or removed as a result of satisfying constraints, but that may or may not be a useful thing to track.

As to the history of M_DUMP_TO_DEEP_OCEAN, it appears to have been in there in the initial commit, and it has been present ever since.  I will guess that means that running in constrained mode has been non-functional since the move from svn.  I verified that it doesn't work in v1.1, but I couldn't try it on any earlier releases because they won't build on my system.
